presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveUnsupportedDynamicFilters.java

```diff
@@ -0,0 +1,154 @@
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.optimizations.PlanOptimizer;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.tree.BooleanLiteral;
+import com.facebook.presto.sql.tree.Expression;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.sql.ExpressionUtils.extractDynamicFilters;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Dynamic filters are supported only right after TableScan
+ */
+public class RemoveUnsupportedDynamicFilters
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        return SimplePlanRewriter.rewriteWith(new RemoveUnsupportedDynamicFilters.Rewriter(), plan, new HashSet<>());
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Set<DynamicFilter>>
+    {
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Set<DynamicFilter>> context)
+        {
+            JoinNode join = (JoinNode) context.defaultRewrite(node, context.get());
+
+            Optional<Expression> filter = join.getFilter();
+            if (filter.isPresent()) {
+                Expression original = filter.get();
+                Expression modified = removeDynamicFilters(original, context.get());
+                if (BooleanLiteral.TRUE_LITERAL.equals(modified)) {
+                    filter = Optional.empty();
+                }
+                else if (!original.equals(modified)) {
+                    filter = Optional.of(modified);
+                }
+            }
+
+            Assignments assignments = join.getDynamicFilterAssignments();
+            Set<DynamicFilter> removedFilters = context.get();
+            if (!removedFilters.isEmpty()) {
+                Assignments originalAssignments = join.getDynamicFilterAssignments();
+                Assignments modifiedAssignments = cleanupAssignments(join.getId(), originalAssignments, removedFilters);
+                if (!originalAssignments.equals(modifiedAssignments)) {
+                    assignments = modifiedAssignments;
+                }
+            }
+
+            if (!filter.equals(join.getFilter()) || !assignments.equals(join.getDynamicFilterAssignments())) {
+                return new JoinNode(
```

Comment:
Just to make it possible to push them through the join node down

So, let's consider the next join tree

```
A
|  -  C
B
|  -  D
E
```

and you want to push dynamic filter that comes from node A to the table scan node E

So, at the first step you should push it down as a filter to node B, and as a second step you have to move it down to the node E. 

But it is possible that you cannot push it down futher ( for example when the input is being generated at the node B itself). So, considering that we do not support dynamic filters at an arbitrary place we need to remove it eventually.